File: backend/app/domain/edit_ops.py
"""改稿结构操作：增删块、改类型。与人手 flex 编辑共用同一套树函数。"""

from __future__ import annotations

from typing import Any
from uuid import uuid4

from pydantic import TypeAdapter

from app.domain.content import Block
from app.domain.flex_edit import default_block_dict, default_text_style_for_type
from app.domain.flex_layout import (
    FlexContainer,
    FlexLeaf,
    find_leaf_parent,
    insert_leaf,
    insert_leaf_after,
    iter_leaf_block_ids,
    prune_empty_containers,
    remove_leaf_by_block_id,
)
from app.domain.flex_normalize import normalize

_block_adapter = TypeAdapter(Block)

STRUCTURAL_TYPES: frozenset[str] = frozenset(
    {"text", "bullets", "image", "chart", "table", "kpi", "cards", "callout"}
)


class EditStructureError(ValueError):
    """结构工具或 apply 无法执行时的可读原因。"""


def parse_block(raw: dict[str, Any]) -> Block:
    return _block_adapter.validate_python(raw)


def dump_block(block: Block) -> dict[str, Any]:
    return block.model_dump(mode="json")


def previous_block_id(tree: FlexContainer | None, block_id: str) -> str | None:
    if tree is None:
        return None
    ids = iter_leaf_block_ids(tree)
    try:
        index = ids.index(block_id)
    except ValueError:
        return None
    if index <= 0:
        return None
    return ids[index - 1]


def add_block(
    blocks: list[Block],
    tree: FlexContainer,
    *,
    block_type: str,
    after_block_id: str,
    content: dict[str, Any] | None = None,
    block_id: str | None = None,
) -> tuple[list[Block], FlexContainer, Block]:
    if block_type not in STRUCTURAL_TYPES:
        raise EditStructureError(f"不支持的块类型 {block_type}")
    if not any(block.id == after_block_id for block in blocks):
        raise EditStructureError(f"锚点块 {after_block_id} 不存在")

    new_id = block_id or uuid4().hex[:12]
    raw = default_block_dict(block_type, new_id)  # type: ignore[arg-type]
    if content:
        raw.update({key: value for key, value in content.items() if key not in {"id", "slot_id"}})
        raw["id"] = new_id
        raw["slot_id"] = new_id
        raw["type"] = block_type
    block = parse_block(raw)

    draft_tree = tree.model_copy(deep=True)
    leaf = FlexLeaf(
        id=f"leaf-{new_id}",
        block_id=new_id,
        text_style=default_text_style_for_type(block_type),  # type: ignore[arg-type]
    )
    if not insert_leaf_after(draft_tree, after_block_id, leaf):
        raise EditStructureError("无法把新块插入布局树")
    normalized = normalize(draft_tree, clamp_title_grow=False)
    return [*blocks, block], normalized, block


def delete_block(
    blocks: list[Block],
    tree: FlexContainer,
    block_id: str,
) -> tuple[list[Block], FlexContainer, Block]:
    target = next((block for block in blocks if block.id == block_id), None)
    if target is None:
        raise EditStructureError(f"内容块 {block_id} 不存在")
    if target.locked:
        raise EditStructureError("该块已人工修改，AI 不会覆盖")

    leaf_ids = iter_leaf_block_ids(tree)
    if block_id in leaf_ids and len(leaf_ids) <= 1:
        raise EditStructureError("至少保留一个内容块")

    draft_tree = tree.model_copy(deep=True)
    if not remove_leaf_by_block_id(draft_tree, block_id):
        raise EditStructureError("布局树中不存在该内容块")
    pruned = prune_empty_containers(draft_tree)
    if not iter_leaf_block_ids(pruned):
        raise EditStructureError("至少保留一个内容块")
    normalized = normalize(pruned, clamp_title_grow=False)
    remaining = [block for block in blocks if block.id != block_id]
    return remaining, normalized, target


def change_block_type(
    blocks: list[Block],
    tree: FlexContainer | None,
    *,
    block_id: str,
    new_type: str,
    content: dict[str, Any] | None = None,
) -> tuple[list[Block], FlexContainer | None, Block, Block]:
    if new_type not in STRUCTURAL_TYPES:
        raise EditStructureError(f"不支持的块类型 {new_type}")
    original = next((block for block in blocks if block.id == block_id), None)
    if original is None:
        raise EditStructureError(f"内容块 {block_id} 不存在")
    if original.locked:
        raise EditStructureError("该块已人工修改，AI 不会覆盖")
    if original.type == new_type:
        raise EditStructureError("块类型没有变化")

    raw = default_block_dict(new_type, block_id)  # type: ignore[arg-type]
    if content:
        raw.update({key: value for key, value in content.items() if key not in {"id", "slot_id"}})
    raw["id"] = block_id
    raw["slot_id"] = original.slot_id
    raw["type"] = new_type
    raw["locked"] = False
    updated = parse_block(raw)

    replaced = [updated if block.id == block_id else block for block in blocks]
    if tree is None:
        return replaced, None, original, updated

    draft_tree = tree.model_copy(deep=True)
    found = find_leaf_parent(draft_tree, block_id)
    if found is not None:
        _parent, _index, leaf = found
        leaf.text_style = default_text_style_for_type(new_type)  # type: ignore[arg-type]
    normalized = normalize(draft_tree, clamp_title_grow=False)
    return replaced, normalized, original, updated


def restore_block(
    blocks: list[Block],
    tree: FlexContainer,
    *,
    block: Block,
    after_block_id: str | None,
) -> tuple[list[Block], FlexContainer]:
    """把删除的块按原位置插回（apply 点「改动前」）。"""
    if any(item.id == block.id for item in blocks):
        return blocks, tree
    draft_tree = tree.model_copy(deep=True)
    leaf = FlexLeaf(
        id=f"leaf-{block.id}",
        block_id=block.id,
        text_style=default_text_style_for_type(block.type)  # type: ignore[arg-type]
        if block.type in STRUCTURAL_TYPES
        else None,
    )
    if after_block_id and any(item.id == after_block_id for item in blocks):
        ok = insert_leaf_after(draft_tree, after_block_id, leaf)
    else:
        ok = insert_leaf(draft_tree, draft_tree.id, 0, leaf)
    if not ok:
        raise EditStructureError("无法把块插回布局树")
    normalized = normalize(draft_tree, clamp_title_grow=False)
    return [*blocks, block], normalized

File: java_backend/src/main/java/com/aippt/domain/content/EditOps.java
package com.aippt.domain.content;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

import com.aippt.domain.flex.FlexContainer;
import com.aippt.domain.flex.FlexEdit;
import com.aippt.domain.flex.FlexLeaf;
import com.aippt.domain.flex.FlexNormalize;
import com.aippt.domain.flex.FlexTrees;

public final class EditOps {

    public static final Set<String> STRUCTURAL_TYPES = Set.of(
            "text", "bullets", "image", "chart", "table", "kpi", "cards", "callout"
    );

    public static class EditStructureException extends RuntimeException {
        public EditStructureException(String message) {
            super(message);
        }
    }

    public record AddResult(List<Map<String, Object>> blocks, FlexContainer tree, Map<String, Object> created) {
    }

    public record DeleteResult(List<Map<String, Object>> blocks, FlexContainer tree, Map<String, Object> removed) {
    }

    public record ChangeResult(
            List<Map<String, Object>> blocks,
            FlexContainer tree,
            Map<String, Object> original,
            Map<String, Object> updated
    ) {
    }

    private EditOps() {
    }

    public static String previousBlockId(FlexContainer tree, String blockId) {
        if (tree == null) {
            return null;
        }
        List<String> ids = FlexTrees.iterLeafBlockIds(tree);
        int index = ids.indexOf(blockId);
        if (index <= 0) {
            return null;
        }
        return ids.get(index - 1);
    }

    public static AddResult add(
            List<Map<String, Object>> blocks,
            FlexContainer tree,
            String blockType,
            String afterBlockId,
            Map<String, Object> content,
            String blockId
    ) {
        if (!STRUCTURAL_TYPES.contains(blockType)) {
            throw new EditStructureException("不支持的块类型 " + blockType);
        }
        if (blocks.stream().noneMatch(block -> afterBlockId.equals(Blocks.id(block)))) {
            throw new EditStructureException("锚点块 " + afterBlockId + " 不存在");
        }
        String newId = blockId == null || blockId.isBlank()
                ? UUID.randomUUID().toString().replace("-", "").substring(0, 12)
                : blockId;
        Map<String, Object> raw = Blocks.defaultBlock(blockType, newId);
        if (content != null) {
            for (Map.Entry<String, Object> entry : content.entrySet()) {
                if (!"id".equals(entry.getKey()) && !"slot_id".equals(entry.getKey())) {
                    raw.put(entry.getKey(), entry.getValue());
                }
            }
            raw.put("id", newId);
            raw.put("slot_id", newId);
            raw.put("type", blockType);
        }
        FlexContainer draft = FlexTrees.copy(tree);
        FlexLeaf leaf = FlexLeaf.of("leaf-" + newId, newId, 1.0, FlexEdit.defaultTextStyle(blockType));
        if (!FlexTrees.insertLeafAfter(draft, afterBlockId, leaf)) {
            throw new EditStructureException("无法把新块插入布局树");
        }
        List<Map<String, Object>> next = new ArrayList<>(blocks);
        next.add(raw);
        return new AddResult(next, FlexNormalize.normalize(draft, false), raw);
    }

    public static DeleteResult delete(List<Map<String, Object>> blocks, FlexContainer tree, String blockId) {
        Map<String, Object> target = blocks.stream()
                .filter(block -> blockId.equals(Blocks.id(block)))
                .findFirst()
                .orElseThrow(() -> new EditStructureException("内容块 " + blockId + " 不存在"));
        if (SlidePatches.locked(target)) {
            throw new EditStructureException("该块已人工修改，AI 不会覆盖");
        }
        List<String> leafIds = FlexTrees.iterLeafBlockIds(tree);
        if (leafIds.contains(blockId) && leafIds.size() <= 1) {
            throw new EditStructureException("至少保留一个内容块");
        }
        FlexContainer draft = FlexTrees.copy(tree);
        if (!FlexTrees.removeLeafByBlockId(draft, blockId)) {
            throw new EditStructureException("布局树中不存在该内容块");
        }
        FlexContainer pruned = FlexTrees.pruneEmptyContainers(draft);
        if (FlexTrees.iterLeafBlockIds(pruned).isEmpty()) {
            throw new EditStructureException("至少保留一个内容块");
        }
        List<Map<String, Object>> remaining = blocks.stream()
                .filter(block -> !blockId.equals(Blocks.id(block)))
                .toList();
        return new DeleteResult(remaining, FlexNormalize.normalize(pruned, false), target);
    }

    public static ChangeResult changeType(
            List<Map<String, Object>> blocks,
            FlexContainer tree,
            String blockId,
            String newType,
            Map<String, Object> content
    ) {
        if (!STRUCTURAL_TYPES.contains(newType)) {
            throw new EditStructureException("不支持的块类型 " + newType);
        }
        Map<String, Object> original = blocks.stream()
                .filter(block -> blockId.equals(Blocks.id(block)))
                .findFirst()
                .orElseThrow(() -> new EditStructureException("内容块 " + blockId + " 不存在"));
        if (SlidePatches.locked(original)) {
            throw new EditStructureException("该块已人工修改，AI 不会覆盖");
        }
        if (newType.equals(Blocks.type(original))) {
            throw new EditStructureException("块类型没有变化");
        }
        Map<String, Object> raw = Blocks.defaultBlock(newType, blockId);
        if (content != null) {
            for (Map.Entry<String, Object> entry : content.entrySet()) {
                if (!"id".equals(entry.getKey()) && !"slot_id".equals(entry.getKey())) {
                    raw.put(entry.getKey(), entry.getValue());
                }
            }
        }
        raw.put("id", blockId);
        raw.put("slot_id", Blocks.slotId(original));
        raw.put("type", newType);
        raw.put("locked", false);
        List<Map<String, Object>> replaced = new ArrayList<>();
        for (Map<String, Object> block : blocks) {
            replaced.add(blockId.equals(Blocks.id(block)) ? raw : block);
        }
        if (tree == null) {
            return new ChangeResult(replaced, null, original, raw);
        }
        FlexContainer draft = FlexTrees.copy(tree);
        FlexTrees.LeafParent found = FlexTrees.findLeafParent(draft, blockId);
        if (found != null) {
            found.parent().children().set(found.index(), found.leaf().withTextStyle(FlexEdit.defaultTextStyle(newType)));
        }
        return new ChangeResult(replaced, FlexNormalize.normalize(draft, false), original, raw);
    }

    public static AddResult restore(
            List<Map<String, Object>> blocks,
            FlexContainer tree,
            Map<String, Object> block,
            String afterBlockId
    ) {
        if (blocks.stream().anyMatch(item -> Blocks.id(block).equals(Blocks.id(item)))) {
            return new AddResult(blocks, tree, block);
        }
        FlexContainer draft = FlexTrees.copy(tree);
        String blockId = Blocks.id(block);
        FlexLeaf leaf = FlexLeaf.of(
                "leaf-" + blockId,
                blockId,
                1.0,
                FlexEdit.defaultTextStyle(Blocks.type(block))
        );
        boolean ok;
        if (afterBlockId != null && blocks.stream().anyMatch(item -> afterBlockId.equals(Blocks.id(item)))) {
            ok = FlexTrees.insertLeafAfter(draft, afterBlockId, leaf);
        } else {
            ok = FlexTrees.insertLeaf(draft, draft.id(), 0, leaf);
        }
        if (!ok) {
            throw new EditStructureException("无法把块插回布局树");
        }
        List<Map<String, Object>> next = new ArrayList<>(blocks);
        next.add(block);
        return new AddResult(next, FlexNormalize.normalize(draft, false), block);
    }
}
